fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryManager.java

```diff
@@ -73,6 +87,70 @@ public synchronized void close() throws IOException {
         createdDirectories.clear();
     }
 
+    @SuppressWarnings("PMD.CloseResource")
+    public CompletableFuture<Void> mergeIndex(LuceneAnalyzerWrapper analyzerWrapper) {
+
+        final ScanProperties scanProperties = ScanProperties.FORWARD_SCAN.with(
+                props -> props.clearState().setReturnedRowLimit(1));
+        final Range range = state.indexSubspace.range();
+        final KeyRange keyRange = new KeyRange(range.begin, range.end);
+
+        final RecordCursor<Tuple> cursor = new ChainedCursor<>(
+                state.context,
+                lastKey -> nextTuple(state.context, state.indexSubspace, keyRange, lastKey, scanProperties),
+                Tuple::pack,
+                Tuple::fromBytes,
+                null,
+                scanProperties);
+
+        return cursor.forEach(groupingKey -> mergeIndex(analyzerWrapper, groupingKey));
+    }
+
+    private void mergeIndex(LuceneAnalyzerWrapper analyzerWrapper, Tuple groupingKey) {
+        try {
+            getDirectoryWrapper(groupingKey).mergeIndex(analyzerWrapper);
+        } catch (IOException e) {
+            if (LOG.isWarnEnabled()) {
+                LOG.warn("Lucene mergeIndex failed: error={}", e.getMessage());
+            }
+        }
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    private CompletableFuture<Optional<Tuple>> nextTuple(@Nonnull FDBRecordContext context,
+                                                         @Nonnull Subspace subspace,
+                                                         @Nonnull KeyRange range,
+                                                         @Nonnull Optional<Tuple> lastTuple,
+                                                         @Nonnull ScanProperties scanProperties) {
+        final KeyValueCursor cursor =
+                lastTuple.isEmpty()
+                ?
+                KeyValueCursor.Builder.withSubspace(subspace)
+                        .setContext(context)
+                        .setRange(range)
+                        .setContinuation(null)
+                        .setScanProperties(scanProperties)
+                        .build()
+                :
+                KeyValueCursor.Builder.withSubspace(subspace)
+                        .setContext(context)
+                        .setContinuation(null)
+                        .setScanProperties(scanProperties)
+                        .setLow(subspace.pack(lastTuple.get().get(0)), EndpointType.RANGE_EXCLUSIVE)
+                        .setHigh(range.getHighKey(), range.getHighEndpoint())
+                        .build();
+
+        return cursor.onNext().thenApply(next -> {
+            if (next.hasNext()) {
+                KeyValue kv = next.get();
+                if (kv != null) {
+                    return Optional.of(subspace.unpack(kv.getKey()));
```

Comment:
This doesn't look right, I think this will pick up all of the key after the subspace, not just the grouping key. So you'll be using whatever the first key used in the index is.

Reply:
✅

Reply:
Would the code below do the trick? 
```
    cursorBuilder.setLow(subspace.pack(Tuple.fromItems(tuple.getItems().subList(0, groupingCount))), EndpointType.RANGE_EXCLUSIVE)
```

(regardless of creating a low level test)

Reply:
Yes, I think that would be the correct code.

Reply:
✅

Reply:
```
LucenOnlineIndexingTest > luceneOnlineIndexingTestGroupingKeys(int) > com.apple.foundationdb.record.lucene.LucenOnlineIndexingTest.luceneOnlineIndexingTestGroupingKeys(int)[1] STARTED
je: gk:("Text 2, and 0", "Text 3, plus 0", "I am text 4, with 0")
je: gk:("Text 2, and 1", "Text 3, plus 1", "I am text 4, with 1")
je: gk:("Text 2, and 2", "Text 3, plus 2", "I am text 4, with 0")
je: gk:("Text 2, and 3", "Text 3, plus 0", "I am text 4, with 1")
je: gk:("Text 2, and 4", "Text 3, plus 1", "I am text 4, with 0")
je: gk:("Text 2, and 5", "Text 3, plus 2", "I am text 4, with 1")
LucenOnlineIndexingTest > luceneOnlineIndexingTestGroupingKeys(int) > com.apple.foundationdb.record.lucene.LucenOnlineIndexingTest.luceneOnlineIndexingTestGroupingKeys(int)[1] PASSED
LucenOnlineIndexingTest > luceneOnlineIndexingTestGroupingKeys(int) > com.apple.foundationdb.record.lucene.LucenOnlineIndexingTest.luceneOnlineIndexingTestGroupingKeys(int)[2] STARTED
je: gk:("Text 2, and 0", "Text 3, plus 0")
je: gk:("Text 2, and 1", "Text 3, plus 1")
je: gk:("Text 2, and 2", "Text 3, plus 2")
je: gk:("Text 2, and 3", "Text 3, plus 0")
je: gk:("Text 2, and 4", "Text 3, plus 1")
je: gk:("Text 2, and 5", "Text 3, plus 2")
LucenOnlineIndexingTest > luceneOnlineIndexingTestGroupingKeys(int) > com.apple.foundationdb.record.lucene.LucenOnlineIndexingTest.luceneOnlineIndexingTestGroupingKeys(int)[2] PASSED
LucenOnlineIndexingTest > luceneOnlineIndexingTestGroupingKeys(int) > com.apple.foundationdb.record.lucene.LucenOnlineIndexingTest.luceneOnlineIndexingTestGroupingKeys(int)[3] STARTED
je: gk:("Text 2, and 0")
je: gk:("Text 2, and 1")
je: gk:("Text 2, and 2")
je: gk:("Text 2, and 3")
je: gk:("Text 2, and 4")
je: gk:("Text 2, and 5")
LucenOnlineIndexingTest > luceneOnlineIndexingTestGroupingKeys(int) > com.apple.foundationdb.record.lucene.LucenOnlineIndexingTest.luceneOnlineIndexingTestGroupingKeys(int)[3] PASSED
```